server/src/test/java/org/elasticsearch/snapshots/SnapshotsServiceTests.java

```diff
@@ -316,23 +367,47 @@ public TestClusterNode currentMaster(ClusterState state) {
 
         private final AllocationService allocationService;
 
-        private final AtomicReference<ClusterState> currentState = new AtomicReference<>();
-
         private final NodeClient client;
 
         private final NodeEnvironment nodeEnv;
 
         private final DisruptableMockTransport mockTransport;
 
+        private final ThreadPool threadPool;
+
+        private Coordinator coordinator;
+
         TestClusterNode(DiscoveryNode node) throws IOException {
             this.node = node;
             final Environment environment = createEnvironment(node.getName());
             masterService = new FakeThreadPoolMasterService(node.getName(), "test", deterministicTaskQueue::scheduleNow);
             final Settings settings = environment.settings();
-            allocationService = ESAllocationTestCase.createAllocationService(settings);
             final ClusterSettings clusterSettings = new ClusterSettings(settings, ClusterSettings.BUILT_IN_CLUSTER_SETTINGS);
-            final ThreadPool threadPool = deterministicTaskQueue.getThreadPool();
-            clusterService = new ClusterService(settings, clusterSettings, threadPool, masterService);
+            threadPool = deterministicTaskQueue.getThreadPool();
+            final Deque<Runnable> orderedTasks = new ArrayDeque<>();
+            clusterService = new ClusterService(settings, clusterSettings, threadPool, masterService,
+                () -> new PrioritizedEsThreadPoolExecutor(node.getName(), 1, 1, 1, TimeUnit.SECONDS,
```

Comment:
I wonder if it's possible to just mock out the thread factory but otherwise reuse the full ESThreadPoolExecutor?

Reply:
Can't do that, unfortunately. I just tried it and what happens is that the normal `execute` delegates to `java.util.concurrent.ThreadPoolExecutor#execute` and that method eventually leads to running `java.util.concurrent.ThreadPoolExecutor#runWorker` on the thread created by the mocked out factory and that method blocks => tests get stuck :(

Reply:
Try the following patch, which will do the trick I think (I added a basic test in `DeterministicTaskQueueTests` as well):

```
diff --git a/server/src/test/java/org/elasticsearch/snapshots/SnapshotsServiceTests.java b/server/src/test/java/org/elasticsearch/snapshots/SnapshotsServiceTests.java
index d7dc4c6541c..2efa2ca9083 100644
--- a/server/src/test/java/org/elasticsearch/snapshots/SnapshotsServiceTests.java
+++ b/server/src/test/java/org/elasticsearch/snapshots/SnapshotsServiceTests.java
@@ -66,7 +66,6 @@ import org.elasticsearch.common.io.stream.NamedWriteableRegistry;
 import org.elasticsearch.common.settings.ClusterSettings;
 import org.elasticsearch.common.settings.IndexScopedSettings;
 import org.elasticsearch.common.settings.Settings;
-import org.elasticsearch.common.unit.TimeValue;
 import org.elasticsearch.common.util.BigArrays;
 import org.elasticsearch.common.util.PageCacheRecycler;
 import org.elasticsearch.common.util.concurrent.PrioritizedEsThreadPoolExecutor;
@@ -104,10 +103,8 @@ import org.junit.Before;
 import java.io.IOException;
 import java.io.UncheckedIOException;
 import java.nio.file.Path;
-import java.util.ArrayDeque;
 import java.util.Collection;
 import java.util.Collections;
-import java.util.Deque;
 import java.util.HashMap;
 import java.util.LinkedHashMap;
 import java.util.List;
@@ -384,29 +381,30 @@ public class SnapshotsServiceTests extends ESTestCase {
             final Settings settings = environment.settings();
             final ClusterSettings clusterSettings = new ClusterSettings(settings, ClusterSettings.BUILT_IN_CLUSTER_SETTINGS);
             threadPool = deterministicTaskQueue.getThreadPool();
-            final Deque<Runnable> orderedTasks = new ArrayDeque<>();
+            class KillWorkerError extends Error {}
             clusterService = new ClusterService(settings, clusterSettings, threadPool, masterService,
-                () -> new PrioritizedEsThreadPoolExecutor(node.getName(), 1, 1, 1, TimeUnit.SECONDS,
+                () -> new PrioritizedEsThreadPoolExecutor(node.getName(), 0, 1, 0, TimeUnit.SECONDS,
                     r -> new Thread() {
                         @Override
                         public void start() {
-                            throw new UnsupportedOperationException();
+                            deterministicTaskQueue.scheduleNow(() -> {
+                                try {
+                                    r.run();
+                                } catch (KillWorkerError kwe) {
+                                    // hacks everywhere
+                                }
+                            });
                         }
                     },
-                    null, null) {
+                    threadPool.getThreadContext(), threadPool.scheduler()) {
 
                     @Override
-                    public void execute(Runnable command, TimeValue timeout, Runnable timeoutCallback) {
-                        throw new UnsupportedOperationException();
+                    protected void afterExecute(Runnable r, Throwable t) {
+                        super.afterExecute(r, t);
+                        // kill worker so that next one will be scheduled
+                        throw new KillWorkerError();
                     }
 
-                    @Override
-                    public void execute(Runnable command) {
-                        // Ensure ordered execution of the tasks here since the threadpool we're
-                        // mocking out is single threaded
-                        orderedTasks.addLast(command);
-                        deterministicTaskQueue.scheduleNow(() -> orderedTasks.pollFirst().run());
-                    }
                 });
             mockTransport = new DisruptableMockTransport(logger) {
                 @Override
diff --git a/test/framework/src/test/java/org/elasticsearch/cluster/coordination/DeterministicTaskQueueTests.java b/test/framework/src/test/java/org/elasticsearch/cluster/coordination/DeterministicTaskQueueTests.java
index 5238cfe8ecd..29e03f37b2e 100644
--- a/test/framework/src/test/java/org/elasticsearch/cluster/coordination/DeterministicTaskQueueTests.java
+++ b/test/framework/src/test/java/org/elasticsearch/cluster/coordination/DeterministicTaskQueueTests.java
@@ -19,8 +19,11 @@
 
 package org.elasticsearch.cluster.coordination;
 
+import org.elasticsearch.common.Priority;
 import org.elasticsearch.common.settings.Settings;
 import org.elasticsearch.common.unit.TimeValue;
+import org.elasticsearch.common.util.concurrent.PrioritizedEsThreadPoolExecutor;
+import org.elasticsearch.common.util.concurrent.PrioritizedRunnable;
 import org.elasticsearch.test.ESTestCase;
 import org.elasticsearch.threadpool.Scheduler;
 import org.elasticsearch.threadpool.ThreadPool;
@@ -30,6 +33,7 @@ import java.util.List;
 import java.util.Random;
 import java.util.concurrent.ExecutorService;
 import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
 import java.util.concurrent.atomic.AtomicBoolean;
 import java.util.concurrent.atomic.AtomicInteger;
 
@@ -422,6 +426,60 @@ public class DeterministicTaskQueueTests extends ESTestCase {
         assertThat(strings, contains("periodic-0", "periodic-1", "periodic-2"));
     }
 
+    public void testPrioritizedEsThreadPoolExecutor() {
+        final DeterministicTaskQueue taskQueue = newTaskQueue();
+        class KillWorkerError extends Error {}
+        final PrioritizedEsThreadPoolExecutor executor = new PrioritizedEsThreadPoolExecutor("test", 0, 1, 0, TimeUnit.MILLISECONDS,
+            r -> new Thread() {
+                @Override
+                public void start() {
+                    taskQueue.scheduleNow(() -> {
+                        try {
+                            r.run();
+                        } catch (KillWorkerError kwe) {
+                            // hacks everywhere
+                        }
+                    });
+                }
+            },
+            taskQueue.getThreadPool().getThreadContext(), taskQueue.getThreadPool().scheduler()) {
+
+            @Override
+            protected void afterExecute(Runnable r, Throwable t) {
+                super.afterExecute(r, t);
+                // kill worker so that next one will be scheduled
+                throw new KillWorkerError();
+            }
+
+        };
+        final AtomicBoolean called1 = new AtomicBoolean();
+        final AtomicBoolean called2 = new AtomicBoolean();
+        executor.execute(new PrioritizedRunnable(Priority.NORMAL) {
+            @Override
+            public void run() {
+                assertFalse(called1.get()); // check that this is only called once
+                called1.set(true);
+            }
+        });
+        executor.execute(new PrioritizedRunnable(Priority.HIGH) {
+            @Override
+            public void run() {
+                assertFalse(called2.get()); // check that this is only called once
+                called2.set(true);
+            }
+        });
+        assertFalse(called1.get());
+        assertFalse(called2.get());
+        taskQueue.runRandomTask();
+        assertFalse(called1.get());
+        assertTrue(called2.get());
+        taskQueue.runRandomTask();
+        assertTrue(called1.get());
+        assertTrue(called2.get());
+        taskQueue.runRandomTask();
+        assertFalse(taskQueue.hasRunnableTasks());
+    }
+
     private static DeterministicTaskQueue newTaskQueue() {
         return newTaskQueue(random());
     }
```

Reply:
@ywelsch hmm two points here:

1. Less important: Is this really making things easier/closer to the real thing?
2. More important: How does this ensure correct order of task execution? It seems we only get correct order of task-enqueuing here?

Reply:
Can you give an example of what you mean by 2? i.e. expected behavior vs actual behavior.

Reply:
@ywelsch nevermind, that was just me seeing ghosts (I could swear I saw the `r` in `Thread#start` being a `TieBreakingRunnable` randomly but can't reproduce that anymore). => Sorry for the noise :)

Should I extract that implementation to its own class maybe so that we have short tests or do you want to?